### PR TITLE
Code review comments

### PR DESCRIPTION
**Pull request by wlevine**
I just put my comments directly in with the code and am making this pull request so that you can see them.

Overall, I think your project look good, and I'm sure you'll continue to make good progress thru the rest of the summer.

Your blog posts were very good for helping me understand your project, thanks for writing those. One thing I still don't get is the relationship between $\theta$ and $\Sigma_\theta$.

A lot of your code I didn't get a chance to read. The review was slow-going for me as there was a lot I didn't understand and had to learn (meaning I didn't understand the statistical model, not that I didn't understand your code). I hope the comments that I did make are useful.

Also, good job documenting your functions.

### CODE REVIEW

File: lib/MixedModels/LMM.rb [hunk 1]
```diff
@@ -149,6 +155,8 @@ def initialize(x:, y:, zt:, lambdat:, weights: nil, offset: 0.0, reml: true,
   #                      are on the right. 
   # * +data+           - a Daru::DataFrame object, containing the response, fixed and random 
   #                      effects, as well as the grouping variables
+  #Maybe the following parameters which appear in all three of these functions should
+  #be encapsulated in a FitOptions class. On the hand, maybe not.
```
Comment (agisga) [inline lib/MixedModels/LMM.rb]: My plan is to implement alternative optimization algorithms, or somehow interface a well-tested minimization library written in C (like NLopt). Then I would do something like
`..., minimizer: :nelder_mead, **minimizer_options)`,
where `**minimizer_options` would collect `start_point:`, `epsilon:`, `max_iterations:` or other optimization options in a Hash (if I remember this syntax correctly).
But maybe an encapsulating class would be better.

File: lib/MixedModels/LMM.rb [hunk 2]
```diff
@@ -165,6 +173,16 @@ def initialize(x:, y:, zt:, lambdat:, weights: nil, offset: 0.0, reml: true,
   #
   def LMM.from_formula(formula:, data:, weights: nil, offset: 0.0, reml: true, 
                        start_point: nil, epsilon: 1e-6, max_iterations: 1e6)
+    #Just out of curiousity, why don't you support the * and ||? Are you planning
+    #to implement them later? I'm not blaming you if you skipped them to keep
+    #the parser simpler, I'm just curious.
```
Comment (agisga) [inline lib/MixedModels/LMM.rb]: Yes, I may implement those later (as well as / for nested random effects), but I think that the formula parser code would become too unclear if I do it now. I want to bring the existing code in a better form first (like separating it out in a different function for example, as you suggest).

File: lib/MixedModels/LMM.rb [hunk 3]
```diff
@@ -176,12 +194,20 @@ def LMM.from_formula(formula:, data:, weights: nil, offset: 0.0, reml: true,
...
+
     #remove whitespaces
     formula.gsub!(%r{\s+}, "")
     # replace ":" with "*", because the LMMFormula class requires this convention
     formula.gsub!(":", "*")
     # deal with the intercept (LMM#from_daru handles the case when both, intercept and 
     # no_intercept, are included)
+    #It took me a while to figure out these 4 lines. Maybe add some comments.
+    #1+ is equivalent to intercept+
+    #The 1 is optional when specifying the formula?
```
Comment (agisga) [inline lib/MixedModels/LMM.rb]: Yes, the intercept term is usually included in linear models by default (having a linear term without an intercept is a statistically flawed practice in most cases), but it needs to be specified manually in pure intercept models like `(1 | g)` or `y ~ 1`.

File: lib/MixedModels/LMM.rb [hunk 4]
```diff
@@ -194,17 +220,32 @@ def LMM.from_formula(formula:, data:, weights: nil, offset: 0.0, reml: true,
     vars = rhs.split %r{\s*[+|()*]\s*}
     vars.delete("")
     vars.uniq!
+
+    #Do you want to impose any constraints on variable names? Are they allowed
+    #to start with numbers or contain symbols like "}"?
```
Comment (agisga) [inline lib/MixedModels/LMM.rb]: I guess, I should exclude everything that cannot be converted to a symbol. I will look into that

File: lib/MixedModels/LMM.rb [hunk 5]
```diff
@@ -229,6 +275,8 @@ def LMM.from_formula(formula:, data:, weights: nil, offset: 0.0, reml: true,
   # The random effects are specified as an Array of Arrays, where the variables in each
   # sub-Array correspond to a common grouping factor (given in the Array +grouping+) and are 
   # modeled as correlated.
+  #Since elements of random_effects correspond to elements of grouping, maybe you
+  #could combine these two together with a hash or small utility class?
```
Comment (agisga) [inline lib/MixedModels/LMM.rb]: Yes, I think that would be more natural. Maybe something like 
`random_effects: {terms: [[:x, :y], [:z]], grouping: [:v, :w]}`

File: lib/MixedModels/LMM.rb [hunk 6]
```diff
@@ -311,6 +359,7 @@ def LMM.from_daru(response:, fixed_effects:, random_effects:, grouping:, data:,
       # replace the categorical variable name in (non-interaction) random_effects 
       random_effects.each_index do |i|
         ind = random_effects[i].find_index(name)
+        #?:
```
Comment (agisga) [inline lib/MixedModels/LMM.rb]: ?

**Comment by agisga** [comment]
> One thing I still don't get is the relationship between $\theta$ and $\Sigma_\theta$.

\Sigma_\theta is the covariance matrix of the random effects vector b. It is parametrized by \theta, but it is parametrized indirectly, because computationally it is much easier to work with its triangular Cholesky factor \Lambda_theta (i.e. \Sigma = \Lamda* \Lamda^T). And since \Lambda_\theta appears in calculations mostly in a transposed form, it is even more efficient to define \Lambda_\theta^T as a function of \theta, which in my code are `lambdat` parametrized by `thfun`.

For example, in the simplest case, when all random effects are independent and identically distributed, then \Sigma_\theta as well as \Lambda_\theta are diagonal matrices. Then `theta` is an Array of length 1, and `lambdat` is a diagonal matrix with `theta[0]` along its diagonal. Consequently, \Sigma_\theta has `theta[0]**2` along its diagonal (i.e. for each i the random effect b_i has variance `theta[0]**2` and standard deviation `theta[0]`).

File: lib/MixedModels/LMM.rb (hunk 6 above)
Comment (wlevine) [inline lib/MixedModels/LMM.rb]: Haha, sorry I forgot to include this comment, I just thought the notation `[ind..ind]` looked strange

**Pull request closed without merging**